### PR TITLE
Replaced existing browser tests with new Protractor tests

### PR DESCRIPTION
**Pull request by jimmynotjim**
Replaced existing browser tests with new Protractor tests

## Additions

- Adds initial config and tests for Protractor

## Removals

- Those busted browser tests

## Changes

- Adds protractor as a dev-dep (make sure you install through `./setup.sh`)

## Testing

- fetch, install protractor and follow the instructions in the README

## Review

- @imuchnik
- @contolini
- @KimberlyMunoz 
- @anselmbradford 
- @sebworks 
- @dpford 
- @kurtw 

## Preview

screen shot 2015-06-25 at 6 07 49 pm

[Preview this PR without the whitespace changes](?w=0)

## Notes

- Initialized with simple `contact-us` and `the-bureau` tests to get started
- Organized into suites to allow testing for one-off instances (typically browser size but could also be non-js tests)
- Utilizes Page Objects for easily sharing pages among multiple tests
- Not sure if removing the old tests will break Jenkins builds, need input from BEWDs
- Once there's a few more tests and we're more sure of the organization we can add the basic instructions to the project readme and getting started guide.

### CODE REVIEW

File: _tests/browser_tests/page_objects/page_contact-us.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+var ContactUsPage = function() {
+  this.get = function() {
+    browser.get( '/contact-us/' );
```
Comment (imuchnik) [inline _tests/browser_tests/page_objects/page_contact-us.js]: This should use the Page Object Model

File: _tests/browser_tests/page_objects/page_the-bureau.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+var TheBureauPage = function() {
+  this.get = function() {
+    browser.get( '/the-bureau' );
```
Comment (imuchnik) [inline _tests/browser_tests/page_objects/page_the-bureau.js]: Same here. This should use the Page Object Model.

File: _tests/browser_tests/environment.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+module.exports = {
+  // The address of a running selenium server.
+  seleniumAddress:
+    (process.env.SELENIUM_URL || 'http://localhost:4444/wd/hub'),
+
+  // Capabilities to be passed to the webdriver instance.
+  capabilities: {
+    'browserName':
+        (process.env.TEST_BROWSER_NAME || 'chrome'),
```
Comment (imuchnik) [inline _tests/browser_tests/environment.js]: This file is great idea. I am not sure we need both multi-capabilities and capabilities. We should probably define the environments and environment specific variables here.

File: _tests/browser_tests/page_objects/page_the-bureau.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  };
+
+  this.pageTitle = function() { return browser.getTitle() };
```
Comment (imuchnik) [inline _tests/browser_tests/page_objects/page_the-bureau.js]: This line is duplicated at in each page. We should refactor into a shared utility file.

**Comment by imuchnik** [comment]
I am just too excited for words.

:+1:  :heart: :heart_eyes:  :love_letter: :trophy: :100:

File: _tests/browser_tests/environment.js (hunk 3 above)
Comment (jimmynotjim) [inline _tests/browser_tests/environment.js]: Yeah, I picked that up from a tutorial I was reading yesterday. I mostly used it so we could set the default host and use the built in baseUrl param instead of the one I created, which required a full `--params.baseURL` instead of just `--baseUrl`. `browser` isn't available to the config either, so it meant we had to call that param on every Page Object instead of just once in the config. Very cool overall.

File: _tests/browser_tests/page_objects/page_the-bureau.js (hunk 4 above)
Comment (jimmynotjim) [inline _tests/browser_tests/page_objects/page_the-bureau.js]: Good call. Should we keep a `utils.js` at the project root?

Comment (imuchnik) [inline _tests/browser_tests/page_objects/page_the-bureau.js]: Some shared or utils folder will be ok.

Comment (jimmynotjim) [inline _tests/browser_tests/page_objects/page_the-bureau.js]: It appears this can't be moved, everything I try I get an error. Will explore again later.

File: _tests/browser_tests/environment.js (hunk 3 above)
Comment (jimmynotjim) [inline _tests/browser_tests/environment.js]: Removed the duplicate capabilities. What other environment specific variables are there to define?

**Comment by contolini** [comment]
:+1:

**Comment by KimberlyMunoz** [comment]
Going through this right now.

Just be sure to add `npm install -g protractor" to either the setup instructions or setup.sh.

File: _tests/browser_tests/README.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+webdriver-manager start
+
+# Open a new tab and tell protractor to start testing:
+protractor conf.js
+```
+
+If you want to test a server other than your local instance (assuming port 7000 is your local CFGov server) use the `--baseUrl` flag. Ex:
+
+```sh
+protractor conf.js --baseUrl=beta.consumerfinance.gov
```
Comment (KimberlyMunoz) [inline _tests/browser_tests/README.md]: This line didn't work. But it did work when I changed it to `protractor conf.js --baseUrl=http://beta.consumerfinance.gov`

**Comment by KimberlyMunoz** [comment]
A couple comments above, but :+1:.

**Comment by KimberlyMunoz** [comment]
All good now.

File: _tests/browser_tests/spec_suites/large_screen/large_contact-us.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    page.get();
+  } );
+
+  it( 'should properly load in a browser', function() {
+    expect( page.pageTitle() ).toBe( 'Contact us' );
+  } );
+
+  it( 'should include 48 individual offices in alpha order', function() {
+    var firstOffice = page.offices.first().element( by.css( '.expandable_label' ) );
+    var lastOffice = page.offices.last().element( by.css( '.expandable_label' ) );
```
Comment (imuchnik) [inline _tests/browser_tests/spec_suites/large_screen/large_contact-us.js]: Ideally all code that is doing some kind of element selection should be in Page Object. The specs should read like a book and contain any lower level concerns of implementation details In this case your spec "knows" the css class of the expandable.

File: _tests/browser_tests/spec_suites/shared/shared_the-bureau.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+TheBureauPage = require( '../../page_objects/page_the-bureau.js' );
```
Comment (anselmbradford) [inline _tests/browser_tests/spec_suites/shared/shared_the-bureau.js]: Is `page_` prefix necessary? Will any other type of js file appear in the `/page_objects/ directory?

**Comment by imuchnik** [comment]
Tiny comments. Nothing that cannot be addressed later, just for good discipline, hygiene and form, worth mentioning now.  
Let  me know if it does not make sense to you.

File: _tests/browser_tests/README.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
+## Quick start:
+
+_Note: If you haven't, make sure to run `./setup.sh from the project root` to install Protractor globally._
+
+From the project root:
+
+```sh
+# Update Selenium Server binaries:
+webdriver-manager update
```
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Does this have to be run every time?

**Comment by imuchnik** [comment]
GREAT JOB!!!!

File: _tests/browser_tests/spec_suites/shared/shared_the-bureau.js (hunk 7 above)
Comment (imuchnik) [inline _tests/browser_tests/spec_suites/shared/shared_the-bureau.js]: @anselmbradford This is purely for clarity and reducing mental load. Without the prefix, you would end up with two files named the same differentiated only by its location in the file system. We felt it would be less confusing to prefix it with 'page_', but everything is game for discussion.

File: _tests/browser_tests/README.md (hunk 8 above)
Comment (imuchnik) [inline _tests/browser_tests/README.md]: No, but might be a good idea.

File: _tests/browser_tests/README.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+```sh
+protractor conf.js --baseUrl=http://beta.consumerfinance.gov
+```
+
+## Pages
+
+Pages are organized in `page_objects`. To add new tests start with adding a new page file and include the elements you'll want to query the page for. For example, if you wanted to test the page title and all the `h2` elements on the about page you'd write:
+
+```js
+var TheBureauPage = function() {
```
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Presently in the project JS we're favoring function declarations over expressions (`function TheBureauPage()` vs `var TheBureauPage = function()`). Declarations get hoisted to the top of the execution context—some say that makes the code misleading. Expressions create a variable that references an anonymous function—some say that makes it harder to debug. At any rate, making the style consistent between project and test code may be preferable.

File: _tests/browser_tests/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+webdriver-manager start
+
+# Open a new tab and tell protractor to start testing:
+protractor conf.js
+```
+
+If you want to test a server other than your local instance (assuming port 7000 is your local CFGov server) use the `--baseUrl` flag. Ex:
+
+```sh
+protractor conf.js --baseUrl=http://beta.consumerfinance.gov
```
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Since the port is mentioned on L:22 perhaps add a port number example here. I run on 7001 so took a moment to realize `protractor conf.js --baseUrl=http://localhost:7001` was needed.

File: _tests/browser_tests/README.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+module.exports = TheBureauPage;
+```
+
+## Tests
+
+Tests are organized into suites. Common tests are in the `/shared` directory and any non-standard test should be added to an existing additional suite or placed into a new suite directory. An example shared suite for our the-bureau example page above would be:
+
+```js
+TheBureauPage = require( '../../page_objects/page_the-bureau.js' );
```
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Does it matter that this is in the global context or should it be `var TheBureauPage`?

File: _tests/browser_tests/README.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+describe( 'Beta The Bureau Page', function() {
+  var page;
+
+  beforeEach( function() {
+    page = new TheBureauPage();
+    page.get();
+  } );
+
+  it( 'should properly load in a browser', function() {
+    expect( page.pageTitle() ).toBe( 'The bureau' );
```
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: This test should fail, no? The title is `The Bureau`.

File: _tests/browser_tests/README.md (hunk 10 above)
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Even better, link to the setting in `environment.js`.

Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Err nm, editing that will mean it gets committed. Should that file be `environment.example.js` and `environment.js` is gitignored?

File: _tests/browser_tests/environment.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+  // Protractor interactive tests
+  interactiveTestPort: 6969,
+
+  // A base URL for your application under test.
+  baseUrl:
+    'http://' + (process.env.HTTP_HOST || 'localhost') +
+          ':' + (process.env.HTTP_PORT || webServerDefaultPort)
+
+};
```
Comment (anselmbradford) [inline _tests/browser_tests/environment.js]: Trivial: keep EOF newlines here and throughout.

File: _tests/browser_tests/page_objects/page_contact-us.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+  };
+
+  this.pageTitle = function() { return browser.getTitle() };
```
Comment (anselmbradford) [inline _tests/browser_tests/page_objects/page_contact-us.js]: Trivial: Missing semicolon `return browser.getTitle();`

File: _tests/browser_tests/page_objects/page_the-bureau.js (hunk 2 above)
Comment (anselmbradford) [inline _tests/browser_tests/page_objects/page_the-bureau.js]: Trivial: some tests include the trailing slash for the URL, others don't. Might make that consistent.

File: _tests/browser_tests/spec_suites/shared/shared_contact-us.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  } );
+
+  it( 'should include the Submit a Complaint phone numbers', function() {
+    var firstPhone = page.complaintPhone.first();
+    var secondPhone = page.complaintPhone.last();
+
+    expect( page.complaintPhone.count() ).toEqual( 2 );
+    expect( page.complaintPhone.getAttribute( 'class' ) ).toMatch( phoneClass );
+    expect( firstPhone.getText() ).toBe( '(855) 411-CFPB (2372)' );
+    expect( firstPhone.getAttribute( 'href' ) ).toBe( 'tel:8554112372' )
```
Comment (anselmbradford) [inline _tests/browser_tests/spec_suites/shared/shared_contact-us.js]: Trivial: Missing ending semicolon.

File: setup.sh [hunk 16]
```diff
@@ -37,6 +37,7 @@ clear(){
 install(){
   echo 'Installing project dependencies...'
   npm install
+  npm install protractor@2.1.0 -g
```
Comment (anselmbradford) [inline setup.sh]: As discussed in chat, ideally the project shouldn't be doing global installs from its setup. Can the server install protractor in the same way it installs grunt, npm, etc. and this installation can go in the Readme? cc @imuchnik @contolini

File: _tests/browser_tests/spec_suites/shared/shared_the-bureau.js (hunk 7 above)
Comment (anselmbradford) [inline _tests/browser_tests/spec_suites/shared/shared_the-bureau.js]: This is versus the `shared_` prefix, got it.

**Comment by anselmbradford** [comment]
Review done. Mostly trivial stuff as you see fit to fix. Doesn't necessarily have to happen here but the main root readme should reference the testing readme instructions. I think ideally we'd have something like `TEST.md` in our root where all the testing instructions are and we eliminate the readmes in the test subdirectories. I feel like readmes in subdirectories can be a little like wikis in that the info can be hard to find when browsing the project.

File: _tests/browser_tests/README.md (hunk 9 above)
Comment (imuchnik) [inline _tests/browser_tests/README.md]: :+1:

File: _tests/browser_tests/README.md (hunk 10 above)
Comment (imuchnik) [inline _tests/browser_tests/README.md]: Not sure I fully understand your concerns here @anselmbradford.

File: _tests/browser_tests/environment.js (hunk 13 above)
Comment (imuchnik) [inline _tests/browser_tests/environment.js]: :+1:

File: _tests/browser_tests/README.md (hunk 10 above)
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: `environment.js` is for setting environment configurations, however for environments that differ from its defaults it can't be modified without being committed—contrary to its purpose. In other projects I've seen this handled by:
- committing `environments.js` with default values (for the server).
- adding `environments.js` to `.gitignore` and committing that.
- adding and committing `environments.example.js` (with default values) and having instructions in the readme to replace `environments.js` with `environments.example.js`.

File: setup.sh (hunk 16 above)
Comment (contolini) [inline setup.sh]: I guess [.travis.yml](https://github.com/cfpb/cfgov-refresh/blob/flapjack/.travis.yml#L24-L25) is the best place for protractor?

File: _tests/browser_tests/README.md (hunk 12 above)
Comment (jimmynotjim) [inline _tests/browser_tests/README.md]: You're correct, it must not check case. I'll still update to match the correct title tag.

File: setup.sh (hunk 16 above)
Comment (anselmbradford) [inline setup.sh]: :+1:

File: _tests/browser_tests/README.md (hunk 10 above)
Comment (jimmynotjim) [inline _tests/browser_tests/README.md]: I'd rather not force travis or jenkins to be writing new files. I've added an example in the readme for changing the port through the passed baseUrl.

File: _tests/browser_tests/README.md (hunk 12 above)
Comment (jimmynotjim) [inline _tests/browser_tests/README.md]: Realized this is in the readme, updated.

**Comment by jimmynotjim** [comment]
@anselmbradford and @imuchnik I made all the recommended changes aside from the readme, I think those can wait so we can move forward for now.

File: _tests/browser_tests/README.md (hunk 10 above)
Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Travis/Jenkins wouldn't be writing files, they would use the committed `environments.js` ... this just allows local repos to override the `environments.js` settings without risking accidentally committing the changes. 

This can be worked out in another PR.

**Comment by anselmbradford** [comment]
Minor readme update, otherwise looks :+1:

File: _tests/browser_tests/README.md (hunk 10 above)
Comment (jimmynotjim) [inline _tests/browser_tests/README.md]: If we use an `environments.example.js` file, we risk people being out of date with the real one. If we don't commit it, Travis/Jenkins have to write the example as to `environments.js`. I don't think either of these scenarios is a good one.

I really think assuming the default local url we already call out in the [Project Readme](https://github.com/cfpb/cfgov-refresh#3-launch-sheer-to-serve-the-site) with an example of how to pass an alternate through the params is the right choice.

Comment (anselmbradford) [inline _tests/browser_tests/README.md]: Okay, just flagging that environments.js is misleading because practically-speaking it can't be changed for local dev environment configurations :stuck_out_tongue_closed_eyes:

File: setup.sh (hunk 16 above)
Comment (imuchnik) [inline setup.sh]: @contolini @anselmbradford well, there are few considerations that may or may not make it a good candidate for Travis. Protractor tests are end-2-end tests which means you need everything. Everything may be a pain in Travis. Alternatively, we can run it in sauce against some other instance, but that would lag behind by at least this commit, which sort  of defeats the purpose. There is one use case that @jimmynotjim  brought up where we can potentially leverage Travis for functional testing.
